[Feature] Add optional custom ID generator option to Mastra instance

nickperry12 opened the pull request:
## Description

Added an option custom ID generator option upon creating an instance of Mastra. Allows users to provide their own ID generator, e.g., ulid, nanoid. If no generator is provided, a fallback of UUID is used.

## Related Issue(s)

Fixes #5958 

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works

nickperry12 committed:
[Feature] Add custom id generator option to Mastra instance

- Added `idGenerator` option to the Mastra class configuration for custom ID generation.
- Provided a fallback of UUID generation if no `idGenerator` option is provided
- Created a test suite to test ID generation with Agent and Memory components
- Fixed a typo in the travel-app package.json file.
- Fixed a test in vNext where the API key was not being correctly accessed

Co-authored-by: Jason Jones <jsunj.jones@gmail.com>

nickperry12 committed:
Reverted some changes in `mastra/index.ts` and agent/index.ts`

Co-authored-by: Jason Jones <jsunj.jones@gmail.com>

nickperry12 committed:
Merge origin/main into feature/replace-id-generator

nickperry12 committed:
Apply feature/replace-id-generator changes after merge

nickperry12 committed:
Refactored tests in the ID generator test suite\n Fixed linting errors\n-Fixed an issue in the workflow tests where the API key wasn't being loaded correctly\n-Fixed a test in workflows that was generating IDs with a mock function, the mock function was not being called correctly

nickperry12 committed:
Added changeset

nickperry12 committed:
Added a new type to ensure the ID generator function does not produce empty strings. An error is now thrown if the ID generator produces an empty string. Refactor the tests appropriately.

nickperry12 commented on packages/core/src/mastra/idgenerator.test.ts:
```diff
@@ -0,0 +1,505 @@
+      const id = mastra.generateId();
+      expect(id).toBe('');
+      expect(emptyIdGenerator).toHaveBeenCalledTimes(1);
+    });
```
@greptileai refactored the tests for empty string handling, it now expects an error to be thrown. Also updated the main implementation -- added a type for the idgenerator function, it now expects the function to return a non-empty string.

nickperry12 committed:
Changed parameter type for setIdGenerator

DanielSLew reviewed (changes requested):
Looks good! Just a minor nit for code cleanup

nickperry12 committed:
Update idgenerator.test.ts

- Fixed linting error in `idgenerator.test.ts`

Co-Authored-By: Jason Jones <jasonjabarjones@gmail.com>

DanielSLew committed:
Merge branch 'main' into feature/replace-id-generator

nickperry12 committed:
Merge branch 'main' into feature/replace-id-generator

abhiaiyer91 committed:
Merge branch 'main' into feature/replace-id-generator

DanielSLew committed:
Merge branch 'main' into feature/replace-id-generator

nickperry12 committed:
Fixed syntax error

Added missing closing braces

DanielSLew committed:
Merge branch 'main' into feature/replace-id-generator

nickperry12 committed:
Resolved merge conflict

- Removed deprecated code from `packages/core/src/agent/index.ts`

- Ran the idgenerator test and confirmed everything is still passing

DanielSLew committed:
Merge branch 'main' into feature/replace-id-generator

nickperry12 committed:
Update index.test.ts

Reverting back to previous to resolve diff

nickperry12 committed:
Update workflow.test.ts

nickperry12 committed:
Update index.ts

Fixing bad merge changes

DanielSLew committed:
Merge branch 'main' into feature/replace-id-generator

nickperry12 committed:
Fix linting issues and workflow test

wardpeet commented:
@nickperry12 Can we ad a follow up PR where you get as an argument the type we are generating it for? You might want have a different generation for messages vs runId

nickperry12 commented:
@wardpeet for sure, that's a great idea!
